src/ring/adapter/jetty9/common.clj

```diff
@@ -46,3 +46,48 @@
   ;; need to upgrade to websockets protocol.
   (and status (== 101 status) ws))
 
+(defn get-charset [^Request request]
+  (when-let [content-type (.. request getHeaders (get HttpHeader/CONTENT_TYPE))]
+    (if-let [mime-type-charset (.get MimeTypes/CACHE content-type)]
+      (str mime-type-charset)
+      (MimeTypes/getCharsetFromContentType content-type))))
+
+(defn- get-client-cert [^Request request]
+  ;; TODO: get data from SecureRequestWithTLSData
+  )
+
+
+(defn build-request-map
+  "Create the request map from the Request object."
+  [^Request request]
+  {:server-port        (Request/getLocalPort request)
+   :server-name        (Request/getLocalAddr request)
+   :remote-addr        (Request/getRemoteAddr request)
+   :uri                (when-let [uri (.getHttpURI request)]
+                         (.getPath uri))
+   :query-string       (when-let [uri (.getHttpURI request)]
+                         (.getQuery uri))
+   :scheme             (when-let [uri (.getHttpURI request)]
+                         (keyword (.getScheme uri)))
+   :request-method     (keyword (.toLowerCase (.getMethod request) Locale/ENGLISH))
+   :protocol           (.getProtocol (.getConnectionMetaData request))
+   :headers            (get-headers request)
+   :content-type       (.. request getHeaders (get HttpHeader/CONTENT_TYPE))
+   :content-length     (.. request getHeaders (get HttpHeader/CONTENT_LENGTH))
+   :character-encoding (get-charset request)
+   :ssl-client-cert    (get-client-cert request)
+   :body               (ContentSourceInputStream. request)})
```

Comment:
you could also have done 

``` (Request/asInputStream request) ```
here although i'm not sure its any different.

Reply:
It should be the same `ContentSourceInputStream` but using `Request/asInputStream` is better so we don't need to touch those io packages.